### ISSUE TITLE
bug: ionic browser remove crosswalk 

### ISSUE THREAD

**Issue by Sjerd** [driftyco/ionic-cli#255]
When you install the crosswalk browser to your project this line gets added to the package.json 
cordovaPlugins list
    "./engine/cordova-crosswalk-engine-c0.6.1"
then when you do ionic browser remove crosswalk the line doesn't get removed and the org.crosswalk.engine is still present in my plugin directory. this will cause my next build to fail. 

In my case adding the crosswalk browser also adds android library: android-support-v4, but because there already was an android-support-v13 library "ionic run android" fails with an unexpected-top-level exception Multiple dex files

**Comment by jbavari** [comment]
@Sjerd thanks for the feedback! This is great.

I'm looking into why the `ionic browser remove crosswalk` isnt working. 

I'll patch this up for you shortly, can you get me a few things?

For starters, `ionic info` - can you copy/paste me that.

Could you tell me what else is listed in that package.json file? Or even, `cordova plugin ls`?

Thanks in advance.

**Comment by Sjerd** [comment]
Ionic info returned this: 
OS: Windows 8
Node Version: v0.10.36
Cordova CLI: 4.2.0
Ionic CLI Version: 1.3.10

These are listed in package.json 
"installed_plugins": {
        "com.danielcwilson.plugins.googleanalytics": {
            "PACKAGE_NAME": "nl.platformveilignederland.wealert"
        },
        "com.ionic.keyboard": {
            "PACKAGE_NAME": "nl.platformveilignederland.wealert"
        },
        "com.phonegap.plugins.PushPlugin": {
            "PACKAGE_NAME": "nl.platformveilignederland.wealert"
        },
        "de.appplant.cordova.plugin.background-mode": {
            "PACKAGE_NAME": "nl.platformveilignederland.wealert"
        },
        "de.appplant.cordova.plugin.local-notification": {
            "PACKAGE_NAME": "nl.platformveilignederland.wealert"
        },
        "nl.x-services.plugins.socialsharing": {
            "PACKAGE_NAME": "nl.platformveilignederland.wealert"
        },
        "nl.x-services.plugins.toast": {
            "PACKAGE_NAME": "nl.platformveilignederland.wealert"
        },
        "org.apache.cordova.camera": {
            "PACKAGE_NAME": "nl.platformveilignederland.wealert"
        },
        "org.apache.cordova.file": {
            "PACKAGE_NAME": "nl.platformveilignederland.wealert"
        },
        "org.apache.cordova.file-transfer": {
            "PACKAGE_NAME": "nl.platformveilignederland.wealert"
        },
        "org.apache.cordova.geolocation": {
            "PACKAGE_NAME": "nl.platformveilignederland.wealert"
        },
        "org.apache.cordova.inappbrowser": {
            "PACKAGE_NAME": "nl.platformveilignederland.wealert"
        },
        "org.apache.cordova.media": {
            "PACKAGE_NAME": "nl.platformveilignederland.wealert"
        },
        "org.apache.cordova.splashscreen": {
            "PACKAGE_NAME": "nl.platformveilignederland.wealert"
        },
        "org.apache.cordova.statusbar": {
            "PACKAGE_NAME": "nl.platformveilignederland.wealert"
        }
    },

and this is what ionic plugin ls returns:

com.danielcwilson.plugins.googleanalytics 0.6.0 "Google Universal Analytics Plug
in"
com.ionic.keyboard 1.0.3 "Keyboard"
com.phonegap.plugins.PushPlugin 2.4.0 "PushPlugin"
de.appplant.cordova.plugin.background-mode 0.6.3 "BackgroundMode"
de.appplant.cordova.plugin.local-notification 0.8.0dev "LocalNotification"
nl.x-services.plugins.socialsharing 4.3.3 "SocialSharing"
nl.x-services.plugins.toast 2.0 "Toast"
org.apache.cordova.camera 0.3.5-dev "Camera"
org.apache.cordova.device 0.2.14-dev "Device"
org.apache.cordova.file 1.3.0 "File"
org.apache.cordova.file-transfer 0.4.5 "File Transfer"
org.apache.cordova.geolocation 0.3.11 "Geolocation"
org.apache.cordova.inappbrowser 0.5.3 "InAppBrowser"
org.apache.cordova.media 0.2.15 "Media"
org.apache.cordova.splashscreen 0.3.5 "Splashscreen"
org.apache.cordova.statusbar 0.1.6 "StatusBar"

**Comment by jbavari** [comment]
@Sjerd can you try latest and see if that resolves it?

**Comment by Sjerd** [comment]
I will try in a couple of hours

**Comment by Sjerd** [comment]
Had this error now but i am probably supposed to just upgrade my cordova.(didn't know that) 
`Failed to install 'org.crosswalk.engine':CordovaError: Plugin doesn't support th
is project's cordova-android version. cordova-android: 3.6.4, failed version req
uirement: >=4.0.0-dev
    at checkEngines (C:\Users\Sjoerd\AppData\Roaming\npm\node_modules\cordova\no
de_modules\cordova-lib\src\plugman\install.js:116:29)
    at _fulfilled (C:\Users\Sjoerd\AppData\Roaming\npm\node_modules\cordova\node
_modules\q\q.js:787:54)
    at self.promiseDispatch.done (C:\Users\Sjoerd\AppData\Roaming\npm\node_modul
es\cordova\node_modules\q\q.js:816:30)
    at Promise.promise.promiseDispatch (C:\Users\Sjoerd\AppData\Roaming\npm\node
_modules\cordova\node_modules\q\q.js:749:13)
    at C:\Users\Sjoerd\AppData\Roaming\npm\node_modules\cordova\node_modules\q\q
.js:557:44
    at flush (C:\Users\Sjoerd\AppData\Roaming\npm\node_modules\cordova\node_modu
les\q\q.js:108:17)
    at process._tickCallback (node.js:442:13)
`

I will try again tomorrow after upgrading cordova.

**Comment by Sjerd** [comment]
I updated cordova. but when i run `Ionic platform add android` my cordova gets downgraded to 3.6.4 and then it gives me te same error again See full log here: 

`

C:\>npm install -g cordova
C:\Users\Sjoerd\AppData\Roaming\npm\cordova -> C:\Users\Sjoerd\AppData\Roaming\n
pm\node_modules\cordova\bin\cordova
cordova@4.3.0 C:\Users\Sjoerd\AppData\Roaming\npm\node_modules\cordova
├── underscore@1.7.0
├── q@1.0.1
├── nopt@3.0.1 (abbrev@1.0.5)
└── cordova-lib@4.3.0 (valid-identifier@0.0.1, unorm@1.3.3, osenv@0.1.0, propert
ies-parser@0.2.3, bplist-parser@0.0.6, mime@1.2.11, semver@2.0.11, dep-graph@1.1
.0, shelljs@0.3.0, rc@0.5.2, npmconf@0.1.16, elementtree@0.1.5, xcode@0.6.7, thr
ough2@0.6.3, glob@4.0.6, d8@0.4.4, init-package-json@1.3.0, request@2.47.0, tar@
1.0.2, plist@1.1.0, cordova-js@3.8.0, npm@1.3.4)

C:\>cd websites/wealertbeta

C:\Websites\WeAlertBeta>ionic platform rm android
Adding in default Ionic hooks

C:\Websites\WeAlertBeta>ionic platform add android
Adding in default Ionic hooks
Running command: "C:\Program Files\nodejs\node.exe" C:\Websites\WeAlertBeta\hook
s\before_platform_add\init_directories.js C:\Websites\WeAlertBeta
npm http GET https://registry.npmjs.org/cordova-android/3.7.1
npm http 200 https://registry.npmjs.org/cordova-android/3.7.1
npm http GET https://registry.npmjs.org/cordova-android/-/cordova-android-3.7.1.
tgz
npm http 200 https://registry.npmjs.org/cordova-android/-/cordova-android-3.7.1.
tgz
Adding android project...
Creating Cordova project for the Android platform:
        Path: platforms\android
        Package: nl.platformveilignederland.wealert
        Name: WeAlert
        Android target: android-21
Copying template files...
Project successfully created.
Running command: "C:\Program Files\nodejs\node.exe" C:\Websites\WeAlertBeta\hook
s\before_prepare\02_jshint.js C:\Websites\WeAlertBeta
Linting www/js/app.js
File www/js/app.js has no errors.
-----------------------------------------
Linting www/js/controllers.js
File www/js/controllers.js has no errors.
-----------------------------------------
Linting www/js/ng-cordova.js
File www/js/ng-cordova.js has no errors.
-----------------------------------------
Linting www/js/services.js
File www/js/services.js has no errors.
-----------------------------------------
Running command: "C:\Program Files\nodejs\node.exe" C:\Websites\WeAlertBeta\hook
s\after_prepare\010_add_platform_class.js C:\Websites\WeAlertBeta
add to body class: platform-android
Running command: "C:\Program Files\nodejs\node.exe" C:\Websites\WeAlertBeta\hook
s\after_prepare\020_remove_sass_from_platforms.js C:\Websites\WeAlertBeta
Installing "com.danielcwilson.plugins.googleanalytics" for android
Installing "com.ionic.keyboard" for android
Installing "com.phonegap.plugins.PushPlugin" for android
Installing "de.appplant.cordova.plugin.background-mode" for android
Installing "org.apache.cordova.device" for android
Installing "de.appplant.cordova.plugin.local-notification" for android
Installing "nl.x-services.plugins.socialsharing" for android
Installing "nl.x-services.plugins.toast" for android
Installing "org.apache.cordova.camera" for android
Plugin "org.apache.cordova.device" already installed on android.
Installing "org.apache.cordova.file" for android
Installing "org.apache.cordova.file-transfer" for android
Installing "org.apache.cordova.geolocation" for android
Installing "org.apache.cordova.inappbrowser" for android
Installing "org.apache.cordova.media" for android
Installing "org.apache.cordova.splashscreen" for android
Installing "org.apache.cordova.statusbar" for android
Installing "org.crosswalk.engine" for android
Failed to install 'org.crosswalk.engine':CordovaError: Plugin doesn't support th
is project's cordova-android version. cordova-android: 3.7.1, failed version req
uirement: >=4.0.0-dev
    at checkEngines (C:\Users\Sjoerd\AppData\Roaming\npm\node_modules\cordova\no
de_modules\cordova-lib\src\plugman\install.js:116:29)
    at C:\Users\Sjoerd\AppData\Roaming\npm\node_modules\cordova\node_modules\cor
dova-lib\src\plugman\install.js:289:16
    at _fulfilled (C:\Users\Sjoerd\AppData\Roaming\npm\node_modules\cordova\node
_modules\q\q.js:787:54)
    at self.promiseDispatch.done (C:\Users\Sjoerd\AppData\Roaming\npm\node_modul
es\cordova\node_modules\q\q.js:816:30)
    at Promise.promise.promiseDispatch (C:\Users\Sjoerd\AppData\Roaming\npm\node
_modules\cordova\node_modules\q\q.js:749:13)
    at C:\Users\Sjoerd\AppData\Roaming\npm\node_modules\cordova\node_modules\q\q
.js:557:44
    at flush (C:\Users\Sjoerd\AppData\Roaming\npm\node_modules\cordova\node_modu
les\q\q.js:108:17)
    at process._tickCallback (node.js:442:13)

C:\Websites\WeAlertBeta>
`

**Comment by Sjerd** [comment]
And removing the browser still gives an error 

```

C:\Websites\WeAlertBeta>ionic browser remove crosswalk
Removing crosswalk browser

C:\Websites\WeAlertBeta>ionic platform rm android
Adding in default Ionic hooks

C:\Websites\WeAlertBeta>ionic platform add android
Adding in default Ionic hooks
Running command: "C:\Program Files\nodejs\node.exe" C:\Websites\WeAlertBeta\hook
s\before_platform_add\init_directories.js C:\Websites\WeAlertBeta
Adding android project...
Creating Cordova project for the Android platform:
        Path: platforms\android
        Package: nl.platformveilignederland.wealert
        Name: WeAlert
        Android target: android-21
Copying template files...
Project successfully created.
Running command: "C:\Program Files\nodejs\node.exe" C:\Websites\WeAlertBeta\hook
s\before_prepare\02_jshint.js C:\Websites\WeAlertBeta
Linting www/js/app.js
File www/js/app.js has no errors.
-----------------------------------------
Linting www/js/controllers.js
File www/js/controllers.js has no errors.
-----------------------------------------
Linting www/js/ng-cordova.js
File www/js/ng-cordova.js has no errors.
-----------------------------------------
Linting www/js/services.js
File www/js/services.js has no errors.
-----------------------------------------
Running command: "C:\Program Files\nodejs\node.exe" C:\Websites\WeAlertBeta\hook
s\after_prepare\010_add_platform_class.js C:\Websites\WeAlertBeta
add to body class: platform-android
Running command: "C:\Program Files\nodejs\node.exe" C:\Websites\WeAlertBeta\hook
s\after_prepare\020_remove_sass_from_platforms.js C:\Websites\WeAlertBeta
Installing "com.danielcwilson.plugins.googleanalytics" for android
Installing "com.ionic.keyboard" for android
Installing "com.phonegap.plugins.PushPlugin" for android
Installing "de.appplant.cordova.plugin.background-mode" for android
Installing "org.apache.cordova.device" for android
Installing "de.appplant.cordova.plugin.local-notification" for android
Installing "nl.x-services.plugins.socialsharing" for android
Installing "nl.x-services.plugins.toast" for android
Installing "org.apache.cordova.camera" for android
Plugin "org.apache.cordova.device" already installed on android.
Installing "org.apache.cordova.file" for android
Installing "org.apache.cordova.file-transfer" for android
Installing "org.apache.cordova.geolocation" for android
Installing "org.apache.cordova.inappbrowser" for android
Installing "org.apache.cordova.media" for android
Installing "org.apache.cordova.splashscreen" for android
Installing "org.apache.cordova.statusbar" for android
Installing "org.crosswalk.engine" for android
Failed to install 'org.crosswalk.engine':CordovaError: Plugin doesn't support th
is project's cordova-android version. cordova-android: 3.7.1, failed version req
uirement: >=4.0.0-dev
    at checkEngines (C:\Users\Sjoerd\AppData\Roaming\npm\node_modules\cordova\no
de_modules\cordova-lib\src\plugman\install.js:116:29)
    at C:\Users\Sjoerd\AppData\Roaming\npm\node_modules\cordova\node_modules\cor
dova-lib\src\plugman\install.js:289:16
    at _fulfilled (C:\Users\Sjoerd\AppData\Roaming\npm\node_modules\cordova\node
_modules\q\q.js:787:54)
    at self.promiseDispatch.done (C:\Users\Sjoerd\AppData\Roaming\npm\node_modul
es\cordova\node_modules\q\q.js:816:30)
    at Promise.promise.promiseDispatch (C:\Users\Sjoerd\AppData\Roaming\npm\node
_modules\cordova\node_modules\q\q.js:749:13)
    at C:\Users\Sjoerd\AppData\Roaming\npm\node_modules\cordova\node_modules\q\q
.js:557:44
    at flush (C:\Users\Sjoerd\AppData\Roaming\npm\node_modules\cordova\node_modu
les\q\q.js:108:17)
    at process._tickCallback (node.js:442:13)

C:\Websites\WeAlertBeta>ionic info

Your system information:

OS: Windows 8
Node Version: v0.10.36
Cordova CLI: 4.3.0
Ionic CLI Version: 1.3.18



C:\Websites\WeAlertBeta>
```
@jbavari

**Comment by jbavari** [comment]
@Sjerd a little info about Crosswalk and the error you're having.

Cordova Android is currently released at 3.6.4 - so when you type `ionic platform add android` it defaults using that version.

That being said, the Cordova Crosswalk Engine (the thing that allows us to use Chromium over the default browser) only works with Cordova Android 4.0.0, which is currently unreleased.

To get around this, the Ionic CLI downloads the latest version of Cordova Android 4.0.0-dev and uses that. That happens by downloading a fork that drifty maintains, storing that in your `./engine` directory, and then using this statement to install it:

`ionic platform add ./engine/cordova-android-c0.5.1` 

This way, Cordova Android 4.0 is installed. 

Then, when you install the Cordova Plugin for Crosswalk, it checks for the version of Cordova Android from this entry in `plugin.xml`:

```
<engines>
   <engine name="cordova-android" version=">=4" />
   <engine name="cordova-plugman" version=">=4.2.0" /><!-- needed for gradleReference support --></engines>
```

Thus, when you run just `ionic platform android` (defaulting back to 3.6.4), this plugin is what is causing your command to fail. 

The fix then to keep using regular android is just to uninstall the cordova crosswalk plugin. This command should help with that:

`ionic browser remove crosswalk`
